chart/scripts/post-delete-hook.sh

```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+set -e
+
+succeeded=()
+failed=()
+echo "Uninstalling Rancher resources in the following namespaces:" "$namespaces"
+for namespace in $namespaces; do
+  app_list=$(helm list -n "$namespace" -q)
+  for app in $app_list; do
+    if [[ $app =~ .crd$ ]]; then
+      echo "--- Skip the app [$app] in the namespace [$namespace]"
+      continue
+    fi
+    echo "--- Deleting the app [$app] in the namespace [$namespace]"
+    if [[ ! $(helm uninstall "$app" -n "$namespace") ]]; then
+      failed=("${failed[@]}" "$app")
+      continue
+    fi
+    t=0
+    # by default, wait 120 seconds in total for an app to be uninstalled
+    while [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]] && [[ $t -lt "$timeout" ]]; do
+      echo "waiting 5 seconds for pods of [$app] to be terminated ..."
+      sleep 5
+      t=$((t + 5))
+    done
+    if [[ "$(kubectl get pods --selector app="$app" -n "$namespace" -o json | jq '.items | length')" != "0" ]]; then
+      echo "timeout uninstalling [$app] in the namespace [$namespace]"
+      failed=("${failed[@]}" "$app")
+    else
+      echo "successfully uninstalled [$app] in the namespace [$namespace]"
+      succeeded=("${succeeded[@]}" "$app")
+    fi
+  done
+  # delete the helm operator pods
+  pod_list=$(kubectl get pods -n "$namespace" -o name --ignore-not-found)
+  for pod in $pod_list; do
+    if [[ $pod =~ ^pod\/helm-operation-* ]]; then
+      echo "--- Deleting the pod [$pod] in the namespace [$namespace]"
+      kubectl delete "$pod" -n "$namespace"
+    fi
+  done
+done
+
+echo "------ Summary ------"
+if [[ ${#succeeded[@]} -ne 0 ]]; then
+  echo "Succeeded to uninstall the following apps:" "${succeeded[@]}"
+fi
+
+if [[ ${#failed[@]} -ne 0 ]]; then
+  echo "Failed to uninstall the following apps:" "${failed[@]}"
+  if [[ "$ignoreTimeoutError" == "false" ]]; then
+    exit 1
+  fi
+else
+  echo "Cleanup finished successfully."
+fi
```

Comment:
Any reason why we want to collect them and then output it at the end as opposed to just printing an error statement and continuing in the for loop itself? Seems like collecting and dumping at the end makes the code more complicated to maintain (because Bash arrays aren't easy to look at 😅 ).

Reply:
I'm good with this either way. It actually does print out as it works or fails. This is just summary.